[draft] [skip ci] Run fuzzing as a travis cron job

minafarid opened the pull request:
* Added a `kRandom` fuzzing target that executes a random target.
* When running travis ci cron job, fuzzing will run a random fuzzing target for 1 hour.

minafarid commented:
Options for continuous (daily) fuzzing:
1. Run all fuzzing targets.
2. Select one target to run by using the day of the year as a guide to which target to fuzz (e.g., the day of the year modulo the number of fuzzing targets).
3. select one target (using approach 2 or random) to fuzz for a longer time while run the other fuzzing targets for shorter periods

var-const commented:
So in concept, option 1 is the most straightforward one. On the other hand, Travis has a time limit for a run, and is also a shared resource. Assuming we add more test targets over time, would each of them get a meaningful slice of time? If the answer is no, then it looks like option 3 is superior to option 2, but is also harder to set up. I'd suggest checking how feasible it is to implement and deciding between 2) and 3) based on that.

@rsgowman @wilhuff What are your thoughts on that? The way I see it is:
* option 2 would iterate over fuzzing targets night by night, e.g. on Monday, run serializer tests, on Tuesday, run field path tests, and so on, looping over. That would give reasonable time for fuzzing to construct a lot of inputs, but any regression might take a week or more before fuzz tests get around to it.
* option 3 is like option 2, except it would do a quick run of each target before doing a long run of a particular target, thus allowing to quickly catch any easily reproducible issues.

These are the approaches that come to mind, though I'll be happy with any other suggestions as well.

rsgowman commented:
Assuming test setup is negligible (seconds) compared to the test time, then I'd prefer 1.

i.e. if you assume that we get a 1h time splice per day, and have 7 fuzz tests, then:

With option 1, each test gets 60m/7 per day for a total of 60m/week.
With option 2, each test gets 60m/week
With option 3, each test gets 60m/week on average.

And option 1 seems simpler.

rsgowman commented:
Ah. Well the obvious question would be can we persist the corpus? But if the answer was "yes, it's easy", then I presume we'd already be doing that.

Looking at the oss-fuzz site, they seem to recommend regularly expanding the seed corpus with inputs that touch new parts of the code. So if we regularly did that, then we'd be ok. But if we don't... then yeah, now I understand the issue. :)

Travis lists a couple options for persisting data between jobs that are worth taking a look at if you haven't already:
https://docs.travis-ci.com/user/build-stages#data-persistence-between-stages-and-jobs

Otherwise, I'm now leaning toward option 2.

var-const commented:
IIUC, corpus grows significantly on each run, so simply persisting the whole corpus each time isn't feasible (@minafarid Mina, can you check this statement/give rough numbers?). Selecting interesting new inputs from the new corpus regularly would be a lot of effort (I've seen some projects doing that with inputs that caused crashes, which is feasible but won't significantly affect the situation).

rsgowman commented:
... and I've changed my mind. :)

I (now) don't like option 2, since it might mean any regressions would take up to n days to be discovered. Ideally, we'd run at least a quick scan over every test every night. I guess this means I'm more in favour of option 3. (Do I get a price for picking all-of-the-above?)

minafarid commented:
The corpus can quickly grow to thousands of interesting input files.

As for the data persistence across stages, it seems that travis recommends using external storage like s3.

minafarid commented:
For reference:
Travis time limits for jobs ([source](https://docs.travis-ci.com/user/customizing-the-build/#build-timeouts)):

* When a job produces no log output for 10 minutes.
* When a job on a public repository takes longer than 50 minutes. 
* When a job on a private repository takes longer than 120 minutes.

We fit in the second category.

minafarid commented:
@var-const @rsgowman updates:
* I added `FUZZING_DURATION` to limit the fuzzing time.
* Added a script that is triggered in travis cron jobs.
* The script:
  * Retrieves all fuzzing targets by executing an invalid fuzz testing script and expecting an error message in the format `Available targets { TARGET1 ... TARGETN }`
  * Defines a `ALLOWED_TIME=2400` as total fuzzing time.
  * Selects one fuzzing target as the main target.
  * Runs the main fuzzing target for half of the time and divides the remaining fuzzing duration equally over the rest of the targets.
  * Examines the fuzzing results by parsing the fuzzing output.
  * A successful fuzzing would have the last line starts with `Done`.
  * Returns `0` if fuzzing succeeds and `-1` if it crashes.

rsgowman commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -56,17 +65,39 @@ FuzzingTarget GetFuzzingTarget() {
+      all_keys.push_back(it->first);
+    }
+    // Print error message with all available targets, which must be enclosed in
+    // curly brackets and separated by spaces.
```
First, I'd expand on this slightly. In particular, I'd recommend mentioning that the fuzzing_travis.sh script parses this output, and it therefore needs to remain stable.

Secondly, I don't like this. :)  But I don't have a better suggestion. :(

var-const commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -56,17 +65,39 @@ FuzzingTarget GetFuzzingTarget() {
-  if (fuzzing_target == "NONE") {
+  // Find the FuzzingTarget enum value that corresponds to the provided
+  // FUZZING_TARGET string value.
+  std::unordered_map<std::string, FuzzingTarget>::const_iterator t =
```
* +1 to `auto`, this looks like a classical example of "how C++11 can simplify your code";
* FWIW, I usually use the name `found` to store the results of a `find`;
* I'm slightly in favor of first dealing with the happy case (i.e., Rich's suggestion to reverse the condition). All other things being equal (as it seems to be in this case), it seems best to nest the shorter case and have the larger piece of code non-nested.

var-const commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -56,17 +65,39 @@ FuzzingTarget GetFuzzingTarget() {
-  LOG_WARN("Invalid fuzzing target: %s", fuzzing_target);
-  return FuzzingTarget::kNone;
+
+  // The function 'atoi' returns 0 if the string could not be correctly parsed.
```
Seeing that this number is never used arithmetically and converted back to a string, perhaps you could avoid this conversion altogether -- just read this variable as a string, pass that as an argument to fuzzer, and let its arguments parsing library handle any errors? `atoi` doesn't allow producing useful information for bad inputs anyway.

var-const commented on scripts/fuzzing_travis.sh:
```diff
@@ -0,0 +1,104 @@
+
+# Get day of the year and use it to select a target.
+DOY="$(date +%j)"
+TARGET="$(( ${DOY} % ${FUZZING_TARGETS_COUNT} ))"
```
To add a couple of suggestions, `todays_target` or `long_running_target`?

minafarid commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -16,8 +16,11 @@
 
 #import <Foundation/NSObject.h>
 #include <string>
+#include <unordered_map>
```
OK. Done.

minafarid commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -16,8 +16,11 @@
+#include <unordered_map>
+#include <vector>
 
 #include "LibFuzzer/FuzzerDefs.h"
```
Ok.

minafarid commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -56,19 +64,39 @@ FuzzingTarget GetFuzzingTarget() {
+// Defaults to "0" if the environment variable is empty, which corresponds to
+// running indefinitely.
+std::string GetFuzzingDuration() {
+  const char *fuzzing_duration_env = std::getenv("FUZZING_DURATION");
```
Done.

minafarid commented on Firestore/Example/FuzzTests/FSTFuzzTestsPrincipal.mm:
```diff
@@ -56,19 +64,39 @@ FuzzingTarget GetFuzzingTarget() {
+std::string GetFuzzingDuration() {
+  const char *fuzzing_duration_env = std::getenv("FUZZING_DURATION");
+
+  if (fuzzing_duration_env == nullptr) {
```
Ok.
